block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagFileStorage.scala

```diff
@@ -298,11 +307,23 @@ final class BlockDagFileStorage[F[_]: Concurrent: Sync: Log: RaiseIOError] priva
                                                             BlockSenderIsMalformed(block)
                                                           )
                                                         }
-                deployHashes = deployData(block).map(_.sig).toList
-                _            <- putBlockNumber(block.blockHash, blockNumber(block))
-                _            <- deployIndex.addAll(deployHashes.map(_ -> block.blockHash))
-                _            <- latestMessagesIndex.addAll(newValidatorsWithSenderLatestMessages.toList)
-                _            <- blockMetadataIndex.add(blockMetadata)
+                deployHashes             = deployData(block).map(_.sig).toList
+                sendersSinceLastProposal <- sendersSinceLastProposalIndex.data
+                _                        <- putBlockNumber(block.blockHash, blockNumber(block))
+                _                        <- deployIndex.addAll(deployHashes.map(_ -> block.blockHash))
+                _                        <- latestMessagesIndex.addAll(newValidatorsWithSenderLatestMessages.toList)
+                _                        <- blockMetadataIndex.add(blockMetadata)
+                _                        <- sendersSinceLastProposalIndex.add(block.sender, (block.blockHash, Set.empty))
+                _ <- sendersSinceLastProposal.toList.traverse_ {
+                      case (validator, (_, senders))
+                          if validator == block.sender || senders.contains(block.sender) =>
+                        ().pure[F]
+                      case (validator, (lastProposedBlock, senders)) =>
```

Comment:
Under what conditions does the second case match?

Reply:
The second case matches for every validator apart from the block sender (which is covered by the first case) and apart from the validators which we know have already seen a proposal from the block sender (also covered by the first case).

This piece of code does not matter anymore; I removed this functionality due to redundancy after a short conversation with @KentShikama. We can deduce seen senders since last proposal by operating purely on latest messages and last proposed block justifications (which now happens in `SynchronyConstraintChecker[F]`).

Reply:
Ah, thanks: I missed that the `if` was part of the `case` pattern.